src/coordinator/ts/values.go

```diff
@@ -133,3 +141,51 @@ func NewFixedStepValues(millisPerStep time.Duration, numSteps int, initialValue
 		values:        values,
 	}
 }
+
+// At every resolution timestep, the last sample value is chosen
+func RawPointsToFixedStep(datapoints Datapoints, start time.Time, end time.Time, interval time.Duration) (FixedResolutionMutableValues, error) {
+	if end.Before(start) {
+		return nil, fmt.Errorf("start cannot be after end, start: %v, end: %v", start, end)
+	}
+
+	if interval == 0 {
+		return nil, errors.New("interval cannot be 0")
+	}
+
+	var numSteps int
+	if end == start {
+		numSteps = 1
+	} else {
+		numSteps = int(end.Sub(start) / interval)
+	}
+
+	fixStepValues := newFixedStepValues(interval, numSteps, math.NaN(), start)
+	fixedResIdx := 0
+	dpIdx := 0
+	numPoints := len(datapoints)
+	for t := start; !t.After(end) && fixedResIdx < numSteps; t = t.Add(interval) {
+		// Find first datapoint not before time t
+		for ; dpIdx < numPoints; dpIdx++ {
+			if !datapoints.DatapointAt(dpIdx).Timestamp.Before(t) {
```

Comment:
Using `datapoints.DatapointAt(dpIdx).Timestamp` pretty often, can possibly pull it out e.g.

```golang
var timestamp time.Time
for t := start; !t.After(end) && fixedResIdx < numSteps; t = t.Add(interval) {
 for ; dpIdx < numPoints; dpIdx++ {
  timestamp = datapoints.DatapointAt(dpIdx).Timestamp
  if timestamp.Before(t) {...
```

Reply:
its used twice and as you rightly noted, we would have to define it outside the for loop. It's a bit ugly so i'll leave it as is unless you show a strong inclination